### PR TITLE
Implement SIA-R62

### PR DESCRIPTION
**Pull request by kasperisager**

### CODE REVIEW

**Comment by kasperisager** [comment]
The rule doesn't yet deal with cases where contrast is used as a distinguishing feature, but let's get it through a first review and reevaluate. It also doesn't yet deal with box shadows and borders similarly to R65.

File: packages/alfa-rules/src/sia-r62/rule.ts [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+                collect &&
+                node
+                  .descendants({ flattened: true })
+                  .some(and(isText, isVisible(device)))
+              ) {
+                return yield node;
+              }
+            }
+
+            // Otherwise, if the element is a <p> element with non-link text
```
[the remark this answers is not in the document]
Comment (kasperisager) [reply packages/alfa-rules/src/sia-r62/rule.ts:48]: That definition entirely disregards the precise element being used, yes. The thing is that R62 won't become a direct implementation of the corresponding ACT-R rule as I'm aiming for something much simpler that we can more easily reason about. To this end, the implementation will use the concept of "paragraph of text" similar to other rules, which we're currently assuming is `<p>` elements that contain visible text content.

File: packages/alfa-rules/src/sia-r62/rule.ts [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+const { and, or, not, test } = Predicate;
+
+export default Rule.Atomic.of<Page, Element, Question>({
+  uri: "https://siteimprove.github.io/sanshikan/rules/sia-r62.html",
+  evaluate({ device, document }) {
+    return {
+      applicability() {
+        return visit(document, false);
+
+        function* visit(node: Node, collect: boolean): Iterable<Element> {
```
[the remark this answers is not in the document]
Comment (kasperisager) [reply packages/alfa-rules/src/sia-r62/rule.ts:29]: That's very much on purpose, yes, for similar reasons as https://github.com/Siteimprove/alfa/pull/538#discussion_r521596865.

File: packages/alfa-rules/src/sia-r62/rule.ts [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+          .get();
+
+        return {
+          1: expectation(
+            test(
+              or(
+                isDistinguishable(container, device),
+                and(
+                  isDistinguishable(container, device, Context.hover(target)),
+                  isDistinguishable(container, device, Context.focus(target))
```
[the remark this answers is not in the document]
Comment (kasperisager) [reply packages/alfa-rules/src/sia-r62/rule.ts:82]: I also believe that's a bit of a problem, which is why R62 requires that links be distinguishable on both hover and focus much like https://www.w3.org/WAI/WCAG21/Techniques/general/G183.

File: packages/alfa-rules/src/sia-r62/rule.ts [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+                      "is-distinguishable-on-hover",
+                      "boolean",
+                      target,
+                      `When hovered, can the link be distinguished from the
+                      surrounding text without relying on color perception
+                      alone?`
+                    ).map((isDistinguishableOnHover) =>
+                      expectation(
+                        isDistinguishableOnHover,
+                        () =>
```
[the remark this answers is not in the document]
Comment (kasperisager) [reply packages/alfa-rules/src/sia-r62/rule.ts:110]: As mentioned in https://github.com/Siteimprove/alfa/pull/538#discussion_r521598143, R62 doesn't.

File: packages/alfa-rules/src/sia-r62/rule.ts [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+
+function isDistinguishable(
+  container: Element,
+  device: Device,
+  context?: Context
+): Predicate<Element> {
+  return (element) => {
+    if (
+      test(not(hasOutline(device, context)), container) &&
+      test(hasOutline(device, context), element)
```
[the remark this answers is not in the document]
Comment (kasperisager) [reply packages/alfa-rules/src/sia-r62/rule.ts:252]: It's a different situation though; R62 looks at a link compared to its container, not itself. If the container has an outline or text decoration, so will the link. R62 therefore only accepts these if they are specified for the link and not the container.

File: packages/alfa-rules/src/sia-r62/rule.ts [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+                            the surrounding text without relying on color
+                            perception alone?`
+                          ).map((isDistinguishableOnFocus) =>
+                            expectation(
+                              isDistinguishableOnFocus,
+                              () => Outcomes.IsDistinguishable,
+                              () => Outcomes.IsNotDistinguishable
+                            )
+                          ),
+                        () => Outcomes.IsNotDistinguishable
```
[the remark this answers is not in the document]
Comment (kasperisager) [reply packages/alfa-rules/src/sia-r62/rule.ts:125]: That's a great point!

File: packages/alfa-rules/src/sia-r62/rule.ts [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+    };
+  },
+});
+
+export namespace Outcomes {
+  export const IsDistinguishable = Ok.of(
+    Diagnostic.of(`The link is distinguishable from the surrounding text`)
+  );
+
+  export const IsNotDistinguishable = Err.of(
```
[the remark this answers is not in the document]
Comment (kasperisager) [reply packages/alfa-rules/src/sia-r62/rule.ts:206]: Extended diagnostics should be used here, yes 👍 Until we have clearer requirements from consumers, let's hold off on adding that though.

**Review by Jym77 — APPROVED** [review]

**Pull request merged**